### PR TITLE
Explore: Add switch to restructure logs for better readability

### LINKED ISSUE #18522: Explore: Prettify json log
**What would you like to be added**:
Some way to pretty print the searchable json logs in the explore ui. Either through an option in the configuration menu or a checkbox in the explore view itself to enable it globally for all json data.

**Why is this needed**:
Reading big chunks of json data is hard and its easy to get lost in it. If there was a way to prettify the data it would be much easier to review and find the thing that you are searching for in the log line.

Example:

Non prettified data:
```
{"log":"{\"type\":\"response\",\"@timestamp\":\"2019-08-12T20:06:43Z\",\"tags\":[],\"pid\":1,\"method\":\"get\",\"statusCode\":200,\"req\":{\"url\":\"/app/kibana\",\"method\":\"get\",\"headers\":{\"user-agent\":\"curl/7.29.0\",\"host\":\"localhost:5601\",\"accept\":\"*/*\"},\"remoteAddress\":\"127.0.0.1\",\"userAgent\":\"127.0.0.1\"},\"res\":{\"statusCode\":200,\"responseTime\":45,\"contentLength\":9},\"message\":\"GET /app/kibana 200 45ms - 9.0B\"}\n","stream":"stdout","time":"2019-08-12T20:06:43.183719829Z"}
```

Prettified data:
```
{
  "log": "{\"type\":\"response\",\"@timestamp\":\"2019-08-12T20:06:43Z\",\"tags\":[],\"pid\":1,\"method\":\"get\",\"statusCode\":200,\"req\":{\"url\":\"/app/kibana\",\"method\":\"get\",\"headers\":{\"user-agent\":\"curl/7.29.0\",\"host\":\"localhost:5601\",\"accept\":\"*/*\"},\"remoteAddress\":\"127.0.0.1\",\"userAgent\":\"127.0.0.1\"},\"res\":{\"statusCode\":200,\"responseTime\":45,\"contentLength\":9},\"message\":\"GET /app/kibana 200 45ms - 9.0B\"}\n",
  "stream": "stdout",
  "time": "2019-08-12T20:06:43.183719829Z"
}
```

One of the problems with this is that the log field has the log data with escaped ```"``` characters. This makes the inner json structure harder to prettify as one needs to take that in consideration when it is prettified.

Prettified data with unescaped ```"```:
```
{  
  "log":{  
    "type":"response",
    "@timestamp":"2019-08-12T20:06:43Z",
    "tags":[  

    ],
    "pid":1,
    "method":"get",
    "statusCode":200,
    "req":{  
      "url":"/app/kibana",
      "method":"get",
      "headers":{  
        "user-agent":"curl/7.29.0",
        "host":"localhost:5601",
        "accept":"*/*"
      },
      "remoteAddress":"127.0.0.1",
      "userAgent":"127.0.0.1"
    },
    "res":{  
      "statusCode":200,
      "responseTime":45,
      "contentLength":9
    },
    "message":"GET /app/kibana 200 45ms - 9.0B"
  },
  "stream":"stdout",
  "time":"2019-08-12T20:06:43.183719829Z"
}
```

### LINKED ISSUE #35298: Loki: printf function to format logs not working while "wrap lines" is disabled
**What happened**:
I'm using a formatting like this 
`line_format "{{ printf \"%-20.20s\" .resource_labels_cluster_name}}\t{{printf \"%-5.5s\" .labels_authorization_k8s_io_decision}}\t{{.protoPayload_resourceName}}"`
to align my log fields vertically. The problem is, this formatting only works when "wrap lines" func enabled. 

We've made a conversation about that at grafana forum, I'm opening this issue regarding to it
https://community.grafana.com/t/aligning-log-columns-vertically/47839/4

Thanks!

### PR DESCRIPTION
**Pull request by olbo98**
**What this PR does / why we need it**:
This PR is a continuation of #36273
This PR adds a switch that toggles whether the json logs will be prettified or not. I've received feedback and will be looking into the possibility of prettifying not only json logs but other log messages as well.

Switch toggled **off**:
first_image

Switch toggled **on**:
second_image

### Edit
Log messages that aren't JSON format can now be restructured as well

Switch toggled **off**:
Screenshot 2021-07-05 at 14 00 40

Switch toggled **on**:
Screenshot 2021-07-05 at 14 00 53

**Things left to do**
- [ ] Add user setting to always restructure logs

**Which issue(s) this PR fixes**:

Fixes: #18522, #35298, #32870

**Special notes for your reviewer**:

### CODE REVIEW

File: public/app/features/explore/Logs.tsx
Comment (tskarhed) [inline public/app/features/explore/Logs.tsx:173]: If you're getting type errors from using `event.target`, try `even.currentTarget`, so you don't have to type cast. Also, is there a reason why the `event` object is optional?

Comment (olbo98) [reply public/app/features/explore/Logs.tsx:173]: Oh i don't know actually, i looked at the other "onChange" functions and saw that they had the event as optional so i did the same. Thanks for the tip regarding `event.currentTarget`, will use that instead 👍

Comment (olbo98) [reply public/app/features/explore/Logs.tsx:173]: It seems i'm getting an error when using `event.currentTarget`. Error:`Property 'checked' does not exist on type 'EventTarget & Element'.`. Maybe that's why i need to typecast with `event.target`?

Comment (tskarhed) [reply public/app/features/explore/Logs.tsx:173]: You might need to play around a bit with the typing. Try `React.ChangeEvent<HTMLInputElement>` for the event type instead.

**Comment by olbo98** [comment]
I added some dividers between the different log options
Screenshot 2021-07-05 at 16 35 07
Would love to have your feedback on this

File: packages/grafana-ui/src/components/Logs/LogRowMessage.tsx
Comment (ivanahuckova) [reply packages/grafana-ui/src/components/Logs/LogRowMessage.tsx:133]: It also doesn't work with Highlighter 

https://user-images.githubusercontent.com/30407135/124589882-c5affc80-de5a-11eb-88f4-b43a649d8ba3.mov

File: packages/grafana-ui/src/components/Logs/LogRowMessage.tsx
Comment (ivanahuckova) [inline packages/grafana-ui/src/components/Logs/LogRowMessage.tsx:94]: Just curious, would it break some functionality if we always used `raw` here ? Because if we have ANSI, raw includes ANSI codes. If we don't have ANSI, entry and raw should be the same. https://github.com/grafana/grafana/blob/main/public/app/core/logs_model.ts#L356

File: public/app/features/explore/Logs.tsx
Comment (ivanahuckova) [inline public/app/features/explore/Logs.tsx:317]: This shouldn't be hard coded such as this. 

Regarding UI, you can maybe check in #UX slack channel on their thoughts how could we make it clear which toggle belongs to which label. My thought would be to make spacing between label and toggle smaller. And increase the space between the non-related toggle and label.

**Comment by ivanahuckova** [comment]
I don't think we need `Add user setting to always restructure logs`, as we are saving this preference in local storage (`store.set(SETTINGS_KEYS.prettifyLogMessage, prettifyLogMessage);`) and it is going to be remembered.

File: packages/grafana-ui/src/components/Logs/LogRowMessage.tsx
Comment (ivanahuckova) [reply packages/grafana-ui/src/components/Logs/LogRowMessage.tsx:133]: Also the code is hard to read with all of the ternary operators. This could be made more readable.

Comment (ivanahuckova) [reply packages/grafana-ui/src/components/Logs/LogRowMessage.tsx:133]: Btw the wrap lines issue could get fixed if this https://github.com/grafana/grafana/issues/35298 was fixed

File: go.mod
Comment (olbo98) [reply go.mod:22]: Yeah i was certain i removed that, don't know how that got in there

File: go.sum
Comment (olbo98) [reply go.sum:192]: Yeah i was certain i removed that, don't know how that got in there

File: packages/grafana-ui/src/components/Logs/LogRowMessage.tsx
Comment (olbo98) [reply packages/grafana-ui/src/components/Logs/LogRowMessage.tsx:56]: what is memoizeOne?

File: packages/grafana-ui/src/components/Logs/LogRowMessage.tsx
Comment (olbo98) [reply packages/grafana-ui/src/components/Logs/LogRowMessage.tsx:133]: Haha i didn't even know highlighting existed. Will look into that!

Comment (olbo98) [reply packages/grafana-ui/src/components/Logs/LogRowMessage.tsx:133]: Is highlighting applied with ANSI code?

File: public/app/features/explore/Logs.tsx
Comment (olbo98) [reply public/app/features/explore/Logs.tsx:317]: Thanks for the tip, i'll see if i can ask them!

File: packages/grafana-ui/src/components/Logs/LogRowMessage.tsx
Comment (ivanahuckova) [reply packages/grafana-ui/src/components/Logs/LogRowMessage.tsx:56]: https://github.com/alexreardon/memoize-one

File: packages/grafana-ui/src/components/Logs/LogRowMessage.tsx
Comment (olbo98) [reply packages/grafana-ui/src/components/Logs/LogRowMessage.tsx]: This should work for highlights now, could you please try again?

File: packages/grafana-ui/src/components/Logs/LogRowMessage.tsx
Comment (olbo98) [reply packages/grafana-ui/src/components/Logs/LogRowMessage.tsx]: Yes i changed it to `const restructuredEntry = restructureLog(raw, prettifyLogMessage);`. Should look more clean now!

File: packages/grafana-ui/src/components/Logs/LogRowMessage.tsx
Comment (olbo98) [reply packages/grafana-ui/src/components/Logs/LogRowMessage.tsx]: https://user-images.githubusercontent.com/16028771/124882626-e13d1380-dfd0-11eb-816a-8d628a46b8ba.mov

File: packages/grafana-ui/src/components/Logs/LogRowMessage.tsx
Comment (ivanahuckova) [inline packages/grafana-ui/src/components/Logs/LogRowMessage.tsx:93]: You could memoize this directly in restructureLog - such as here in [sortLogs](https://github.com/grafana/grafana/blob/main/packages/grafana-ui/src/components/Logs/LogRows.tsx#L76-L78)

File: public/app/features/explore/Logs.tsx
Comment (ivanahuckova) [inline public/app/features/explore/Logs.tsx:323]: We should decide what is the naming of this going to be going to be and keep it consistent, cause somewhere you are using prettify and somewhere restructure and ideally the naming would be same. Maybe we can do quick check on standup? 🙂

File: packages/grafana-ui/src/components/Logs/LogRowMessage.tsx
Comment (ivanahuckova) [inline packages/grafana-ui/src/components/Logs/LogRowMessage.tsx:71]: This is not related to your PR nit pick - just fyi - in `needsHighlighter` we could do this (which would make typing easier.
```
Boolean(
      highlights && highlights?.length > 0 && highlights[0] && highlights[0].length > 0 && entry.length < MAX_CHARACTERS
    )
```

**Review by ivanahuckova — COMMENTED** [review]
We are almost at the finish line... 🚀 🤘 Good job on this!

File: public/app/features/explore/Logs.tsx
Comment (olbo98) [reply public/app/features/explore/Logs.tsx:323]: Yes i agree, that's a very good idea! Let's get everyone's opinion on the standup today 😄

File: packages/grafana-ui/src/components/Logs/LogRowMessage.tsx
Comment (olbo98) [reply packages/grafana-ui/src/components/Logs/LogRowMessage.tsx:93]: Oh that's smart! By the way, is it a good idea to memoize the `renderLogMessage` function as well?

Comment (ivanahuckova) [reply packages/grafana-ui/src/components/Logs/LogRowMessage.tsx]: Not necessarily. Memoization is useful when function is pure (which it is in both cases) and when we are doing heavier computing functions - e.g. figuring out parser and then parsing each log line that can be long (which we do in `restrucutreLog`). `renderLogMessage` is pretty simple and just returns components based on arguments. 

So the memoization in `restructureLog` is useful so we don't have to do whole parsing every time we toggle context or change wrapLogMessage. or add a text to highlight.

File: public/app/plugins/panel/logs/LogsPanel.tsx
Comment (ivanahuckova) [inline public/app/plugins/panel/logs/LogsPanel.tsx:12]: In LogsPanel, this is not going to work because you haven't add the switch to control this setting. The switch should be added in `public/app/plugins/panel/logs/module.tsx`
image

File: packages/grafana-ui/src/components/Logs/LogRowMessage.tsx
Comment (ivanahuckova) [inline packages/grafana-ui/src/components/Logs/LogRowMessage.tsx:119]:
```suggestion
```

**Review by ivanahuckova — CHANGES_REQUESTED** [review]

File: public/app/plugins/panel/logs/LogsPanel.tsx
Comment (olbo98) [reply public/app/plugins/panel/logs/LogsPanel.tsx:12]: Oh i see, i had no idea! Thanks for the information 😄

**Review by ivanahuckova — COMMENTED** [review]
This looks great and works as expected. I have 1 last question, which I noticed just now when I tested this with JSON logs. Is there a reason why we don't use `return JSON.stringify(JSON.parse(line), undefined, 2);` as we initially talked about for JSON logs that formats also nested fields - e.g. we have

```
{
  "user": "",
  "request": {
    "time": "0.027",
    "method": "GET",
    "size": "55",
  },
  "response": {
    "status": 200,
    "test": "OK",
  }
}
```

instead of 

```
{
  "user": "",
  "request": {"time": "0.027", "method": "GET", "size": "55"},
  "response": {"status": 200, "test": "OK"}
}
```

**Comment by olbo98** [comment]
> This looks great and works as expected. I have 1 last question, which I noticed just now when I tested this with JSON logs. Is there a reason why we don't use `return JSON.stringify(JSON.parse(line), undefined, 2);` as we initially talked about for JSON logs that formats also nested fields - e.g. we have
> 
> ```
> {
>   "user": "",
>   "request": {
>     "time": "0.027",
>     "method": "GET",
>     "size": "55",
>   },
>   "response": {
>     "status": 200,
>     "test": "OK",
>   }
> }
> ```
> 
> instead of
> 
> ```
> {
>   "user": "",
>   "request": {"time": "0.027", "method": "GET", "size": "55"},
>   "response": {"status": 200, "test": "OK"}
> }
> ```

I saw that `getParser()` returns the correct parser depending on the format of the log which i thought was handy. But if what you're saying is true and we don't get the intended format as we talked about then i will change it 👍

**Comment by olbo98** [comment]
@ivanahuckova @Elfo404 i noticed some "bugs" while testing the perttify function. I generated log lines that seems to break the functionality of prettify somewhat. 
Take for example this log message:
Screenshot 2021-07-14 at 15 25 34
When using prettify this is the result:
Screenshot 2021-07-14 at 15 25 20
As you can see the initial "Info:" part at the beginning of the un-prettified log has been removed. I don't know if this is important or not but it seems a little bit weird.

Secondly, if you look at the "msg" field you can see that part of the original value has been cut off. This seems to be a bug with nested quotation marks but on the other hand is an expected result of the logfmt parser and i'm not sure how to fix this

What are your thoughts?

**Comment by olbo98** [comment]
I've updated prettify to only work on JSON logs now as we discussed! It would be nice if someone @ivanahuckova @Elfo404  could test and see if it works with highlighting

File: public/app/plugins/panel/logs/module.tsx
Comment (ivanahuckova) [inline public/app/plugins/panel/logs/module.tsx:33]:
```suggestion
      name: 'Prettify JSON',
```

**Review by ivanahuckova — APPROVED** [review]
Nice! Works great. LGTM! 🚀

File: public/app/plugins/panel/logs/module.tsx
Comment (olbo98) [reply public/app/plugins/panel/logs/module.tsx:33]: Oh right! I always forget something haha

**Comment by ivanahuckova** [comment]
You still probably need to do `git merge origin main`.

**Comment by olbo98** [comment]
@ivanahuckova should we add some documentation for this PR before merging?

**Comment by ivanahuckova** [comment]
@ivanahuckova should we add some documentation for this PR before merging?

I think you can create separate PR that mentions this one and add @achatterjee-grafana as a reviewer.

**Pull request merged**
